File: core/src/com/github/servb/pph/gxlib/gxlmetrics/Size.java
/*
 * Copyright 2018 SerVB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.servb.pph.gxlib.gxlmetrics;

import org.checkerframework.checker.signedness.qual.Unsigned;

/**
 * The Size class.
 *
 * @author SerVB
 */
public final class Size implements ISize {

    /** Width. */
    private @Unsigned int w;
    /** Height. */
    private @Unsigned int h;

    /**
     * Constructs the object by metrics.
     *
     * @param w The width.
     * @param h The height.
     */
    public Size(final @Unsigned int w, final @Unsigned int h) {
        this.w = w;
        this.h = h;
    }

    /**
     * Constructs the object by the other object.
     *
     * @param other The other size.
     */
    public Size(final IConstSize other) {
        this.w = other.getW();
        this.h = other.getH();
    }

    @Override
    public @Unsigned int getW() {
        return w;
    }

    @Override
    public @Unsigned int getH() {
        return h;
    }

    @Override
    public ISize operatorP(final IConstSize other) {
        return new Size(w + other.getW(), h + other.getH());
    }

    @Override
    public ISize operatorP(final @Unsigned int other) {
        return new Size(w + other, h + other);
    }

    @Override
    public ISize operatorM(final @Unsigned int other) {
        return new Size(w - other, h - other);
    }

    @Override
    public boolean operatorEe(final IConstSize other) {
        return w == other.getW() && h == other.getH();
    }

    @Override
    public float GetAspectRatio() {
        return ((float) w) / ((float) h);
    }

    @Override
    public boolean IsZero() {
        return w == 0 && h == 0;
    }

    @Override
    public void setW(final @Unsigned int value) {
        w = value;
    }

    @Override
    public void setH(final @Unsigned int value) {
        h = value;
    }

    @Override
    public void operatorPe(final @Unsigned int value) {
        w += value;
        h += value;
    }

    @Override
    public void operatorMe(final @Unsigned int value) {
        w -= value;
        h -= value;
    }

    @Override
    public void InflateSize(final @Unsigned int w_offs, final @Unsigned int h_offs) {
        w += w_offs;
        h += h_offs;
    }

    @Override
    public void InflateSize(final @Unsigned int offs) {
        InflateSize(offs, offs);
    }

    @Override
    public void DeflateSize(final @Unsigned int w_offs, final @Unsigned int h_offs) {
        w -= w_offs;
        h -= h_offs;
    }

    @Override
    public void DeflateSize(final @Unsigned int offs) {
        DeflateSize(offs, offs);
    }

    @Override
    public void Zero() {
        w = h = 0;
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 53 * hash + this.w;
        hash = 53 * hash + this.h;
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Size other = (Size) obj;
        if (this.w != other.w) {
            return false;
        }
        return this.h == other.h;
    }
}

File: core/src/main/kotlin/com/github/servb/pph/gxlib/gxlmetrics/Size.kt
package com.github.servb.pph.gxlib.gxlmetrics

/** Immutable (constant) Size view. */
@ExperimentalUnsignedTypes
interface Sizec {
    val w: UInt
    val h: UInt

    /**
     * Returns the aspect ratio width/height.
     *
     * @return The aspect ratio.
     */
    // TODO: Remove toLong
    fun GetAspectRatio() = w.toLong().toFloat() / h.toLong().toFloat()

    /**
     * Checks if the size is equal to zero.
     *
     * @return The result of the check.
     */
    fun IsZero() = w == 0u && h == 0u

    operator fun plus(other: Sizec): Sizec = Size(w + other.w, h + other.h)
    operator fun minus(other: Sizec): Sizec = Size(w - other.w, h - other.h)
    operator fun plus(offs: UInt): Sizec = Size(w + offs, h + offs)
    operator fun minus(offs: UInt): Sizec = Size(w - offs, h - offs)

    fun toSizec(): Sizec = Size(this)
    fun toSize() = Size(this)
}

@ExperimentalUnsignedTypes
data class Size(override var w: UInt, override var h: UInt) : Sizec {
    constructor() : this(0u, 0u)

    constructor(other: Sizec) : this(other.w, other.h)

    operator fun plus(other: Size) = Size(w + other.w, h + other.h)
    operator fun minus(other: Size) = Size(w - other.w, h - other.h)

    /**
     * Inflates the size by the value.
     *
     * @param value The value.
     */
    @Deprecated("Use #InflateSize instead.", ReplaceWith("InflateSize(value)"))
    operator fun plusAssign(value: UInt) = InflateSize(value)

    /**
     * Deflates the size by the value.
     *
     * @param value The value.
     */
    @Deprecated("Use #DeflateSize instead.", ReplaceWith("DeflateSize(value)"))
    operator fun minusAssign(value: UInt) = DeflateSize(value)

    /**
     * Inflates the size by the value.
     *
     * @param w_offs The width offset.
     * @param h_offs The height offset.
     */
    fun InflateSize(w_offs: UInt, h_offs: UInt) {
        w += w_offs
        h += h_offs
    }

    /**
     * Inflates the size by the value.
     *
     * @param offs The value.
     */
    fun InflateSize(offs: UInt) = InflateSize(offs, offs)

    /**
     * Deflates the size by the value.
     *
     * @param w_offs The width offset.
     * @param h_offs The height offset.
     */
    fun DeflateSize(w_offs: UInt, h_offs: UInt) {
        w -= w_offs
        h -= h_offs
    }

    /**
     * Deflates the size by the value.
     *
     * @param offs The value.
     */
    fun DeflateSize(offs: UInt) = DeflateSize(offs, offs)

    /** Sets width and height to zero.  */
    fun Zero() {
        h = 0u
        w = 0u
    }
}
